backupccl: create stripped crdb_internal.fingerprint overload

msbutler opened the pull request:
This patch adds a variant of crdb_internal.fingerprint(), which creates a "stripped" fingerprint of the target span. Namely, `crdb_internal.fingerping(span,true)` will return a fingerprint that is agnostic to the mvcc timestamps and the index prefix of the key, and considers only the latest mvcc history of the key span.

For example, suppose the user fingerprinted a table at some system time, then backed up and restored it to that same system time. The restored table should have the same fingerprint!

This fingerprint variant is signicantly more scalable than SHOW EXPERIMENTAL FINGERPRINT, as it uses export requests compared to a simple scan.

Fixes #98570

Release note: None

msbutler committed:
backupccl: create stripped crdb_internal.fingerprint overload

This patch adds a variant of crdb_internal.fingerprint(), which creates a
"stripped" fingerprint of the target span. Namely,
`crdb_internal.fingerpint(span,true)` will return a fingerprint that is
agnostic to the mvcc timestamps and the index prefix of the key, and considers
only the latest mvcc history of the key span. This cmd should only get used
over user table space, i.e. on keys with an index prefix.

For example, suppose the user fingerprinted a table at some system time, then
backed up and restored it to that same system time. The restored table should
have the same fingerprint!

crdb_internal.fingerpint(span,false) is equivalent to
crdb_internal.fingerpint(span,NULL,LATEST).

This fingerprint variant is signicantly more scalable than SHOW EXPERIMENTAL
FINGERPRINT, as it uses export requests compared to a simple scan.

Fixes #98570

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98759)
<!-- Reviewable:end -->

stevendanna commented on pkg/storage/fingerprint_writer.go:
```diff
@@ -49,8 +51,13 @@ func makeFingerprintWriter(
 	opts MVCCExportFingerprintOptions,
-) fingerprintWriter {
+) (fingerprintWriter, error) {
+	_, tenantPrefix, err := keys.DecodeTenantPrefix(startKey)
+	if err != nil {
+		return fingerprintWriter{}, err
+	}
```
I believe we are decoding this so that we can use StripTenantPrefix below which asserts that the prefix is the same.  But what if I want to use this from the system tenant on a span that crosses tenant boundaries. Are we saying that isn't a valid use case. If so, we should include some reasoning for why not.

msbutler replied:
yeah, this is a good point. Though I wonder if using a codec's `StipTenantPrefix` is more performant than `DecodeTenantPrefix`. I can run some benchmarks.

msbutler replied:
there's no noticeable perf difference between `StripTenantPrefix` vs `DecodeTenantPrefix`, according to a macro benchmark (fingerpinting a 15 gb table). So I suppose I shouldn't add this restriction to limit the cmd to a single tenant key span -- i'll need to create a slightly more complicated helper to strip the index prefix then.

msbutler commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -7491,6 +7487,24 @@ expires until the statement bundle is collected`,
+			ReturnType: tree.FixedReturnType(types.Int),
+			Fn: func(ctx context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
+				if skipTimestamp := *args[1].(*tree.DBool); !skipTimestamp {
+					return nil, errors.New("cannot use crdb_internal.fingerprint skip timestamp overload with skip_ts=false")
```
yeah it would be equivalent. I suppose we could treat `SELECT crdb_internal.fingerprint(span, false)` as syntactic sugar for `SELECT crdb_internal.fingerprint(span, 0::TIMESTAMPTZ, false)`. 

I initially disallowed this form to prevent confusion-- "if you use this first overload, you can only use it for the stripped fingerprint".

I'll change this to allow the sugared syntax.

msbutler committed:
sql/keys introduce key decoders that do not rely on a codec

The codec normally used to decode and encode crdb keys requires a tenant prefix
upon creation. For callers that want to manipulate keys across tenants, the per
tenant codec provides a poor experience. This patch provides a few helper
functions to strip key prefixes without a codec.

Informs #98570
Release note: None

msbutler commented:
unrelated CI flake. RFAL!

yuzefovich reviewed (commented):
I have some drive-by nits, hope you don't mind :smile: 

> Reviewed 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-:-NREY85E59y_S8RKMFOB:b73gaoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @msbutler, and @stevendanna)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREWbyL0XXCQEKVGoEW):*
nit: missing `Release note: None`.
___
*[`pkg/keys/sql.go` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREWeWKE9YjqS79uICQ:-NREWeWKE9YjqS79uICR:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e2a6fd84ad07e13b66662b130a460fc19ae5480/pkg/keys/sql.go#L70)):*
> ```Go
> func StripTenantPrefixE(key roachpb.Key) ([]byte, error) {
> 	var err error
> 	// Strip the tenant prefix
> ```

nit: missing period.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7500 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXl5X003QTl09PVyS:-NREXl5X003QTl09PVyT:be6xbfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7500)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.Int),
> 			Fn: func(ctx context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
> 				skipTimestamp := *args[1].(*tree.DBool)
> ```

nit: it would be good to add an explicit length check on `args` to be 2 here (and 3 in the other overload) as well as to check the types of each arg (i.e. return an error if `args[1]` is not a `DBool`).
___
*[`pkg/sql/sem/builtins/builtins.go` line 7501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXxh78GdNKOFuOxe5:-NREXxh78GdNKOFuOxe6:b-oeaklx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7501)):*
> ```Go
> 			Fn: func(ctx context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
> 				skipTimestamp := *args[1].(*tree.DBool)
> 				return fingerprint(ctx, evalCtx, args, hlc.Timestamp{}, false, bool(skipTimestamp))
> ```

nit: add inlined comment for what `false` stands here for.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7520 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREY0Gy7B4Dfryc7n8x:-NREY0Gy7B4Dfryc7n8y:b-znuu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7520)):*
> ```Go
> 				startTimestamp := hlc.Timestamp{WallTime: startTime.UnixNano()}
> 				allRevisions := *args[2].(*tree.DBool)
> 				return fingerprint(ctx, evalCtx, args, startTimestamp, bool(allRevisions), false)
> ```

nit: ditto for inlined comment for `false`.
___
*[`pkg/sql/sem/builtins/fingerprint_builtins.go` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXbO_BhRxkjO9uhnk:-NREXbO_BhRxkjO9uhnl:b-qwhbpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/fingerprint_builtins.go#L37)):*
> ```Go
> 	ctx context.Context,
> 	evalCtx *eval.Context,
> 	args tree.Datums,
> ```

nit: since we only use `args[0]` and we assume it's a `DArray`, should we just take in `DArray` explicitly here?


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/keys/sql.go:
```diff
@@ -65,6 +65,45 @@ func DecodeTenantPrefixE(key roachpb.Key) ([]byte, roachpb.TenantID, error) {
 	return rem, id, nil
 }
 
+func StripTenantPrefixE(key roachpb.Key) ([]byte, error) {
```
[nit] We only added the `E` postfix to DecodeTenantPrefixE to differentiate it from the version that panics. We can probably avoid that here.

[nit] Might want to add a documentation comment.

stevendanna commented on pkg/storage/fingerprint_writer.go:
```diff
@@ -177,7 +190,15 @@ func (f *fingerprintWriter) stripValueChecksum(value []byte) []byte {
+	return remainder
+}
+
+func (f *fingerprintWriter) stripTablePrefix(key []byte) []byte {
+	remainder, err := keys.StripIndexPrefixE(key)
```
[nit] A little surpising to see StripIndexPrefix as the first call in a function named stripTablePrefix.

stevendanna commented on pkg/storage/testdata/mvcc_histories/export_fingerprint_no_timestamps:
```diff
@@ -0,0 +1,77 @@
+
+
+run ok
+put k=a ts=2 v=a2
```
NB: If you want to test the table prefix stripping as well, if you do something like `put k=/b ts=2 v=b tenant-prefix=10 init-checksum` you'll get a key that looks like a table key.

stevendanna reviewed (commented):
Overall I think will be useful. I've left some nitpicks.

On the various Strip functions: I could go either way. They introduce a bit of duplication, but it is nice to be able to operate on a roachpb.Key without specific codec. If you like them, I'd say keep them.

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98759)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @stevendanna, and @yuzefovich)

___
*[`pkg/sql/sem/builtins/builtins.go` line 7500 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXl5X003QTl09PVyS:-NRJTmtYDQ6dUlDeZmJi:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7500)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it would be good to add an explicit length check on `args` to be 2 here (and 3 in the other overload) as well as to check the types of each arg (i.e. return an error if `args[1]` is not a `DBool`).
</blockquote></details>

done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXxh78GdNKOFuOxe5:-NRJUOSI46RXPrdHQ8fx:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7501)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: add inlined comment for what `false` stands here for.
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/builtins.go` line 7520 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREY0Gy7B4Dfryc7n8x:-NRJUXNREsC19uAOCsIU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/builtins.go#L7520)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto for inlined comment for `false`.
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/fingerprint_builtins.go` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREXbO_BhRxkjO9uhnk:-NRJU_oH2VELt7hDeOKm:b-n0vhv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc09b489949c309bf4a4989b35027fbb3a02a54/pkg/sql/sem/builtins/fingerprint_builtins.go#L37)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: since we only use `args[0]` and we assume it's a `DArray`, should we just take in `DArray` explicitly here?
</blockquote></details>

i refactored this so we parse the span outside fingerprint(). 


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98759)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @stevendanna, and @yuzefovich)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98759#-NREWbyL0XXCQEKVGoEW):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: missing `Release note: None`.
</blockquote></details>

will plan to clean up the commit message once I squash everything!


<!-- Sent from Reviewable.io -->

msbutler commented:
TFTRs!!!

bors r=stevendanna
